core/dom/range.js

```diff
@@ -2818,6 +2818,73 @@ CKEDITOR.dom.range = function( root ) {
 			reference.remove();
 		},
 
+		/**
+		 * Returns array with rectangles of areas covered by ranges. For each DOM element within range there is one rectangle, but it represents only part of element which is within range, not whole element.
+		 *
+		 * @since 4.9.0
+		 * @returns {Array}
+		 */
+		getClientRects: function() {
+			// Extending empty object wth rect, to prevent inheriting from DOMRect, same approach as in CKEDITOR.dom.element.getClientRect().
+			function convertRect( rect ) {
+				var newRect = CKEDITOR.tools.extend( {}, rect );
+				// Some browsers might not return width and height.
+				!newRect.width && ( newRect.width = newRect.right - newRect.left );
+				!newRect.height && ( newRect.height = newRect.bottom - newRect.top );
+				return newRect;
+			}
+
+			// Fallback helper for browsers that don't support native getClientRects()
+			function getRect( context ) {
+				var bookmark = context.createBookmark(),
+					start = bookmark.startNode,
+					end = bookmark.endNode,
+					rects,
+					rect = {};
+
+				start.setText = '';
+				start.removeStyle( 'display' );
+
+				if ( end ) {
+					end.setText = '';
+					end.removeStyle( 'display' );
+					rects = [ start.getClientRect(), end.getClientRect() ];
+					start.remove();
+					end.remove();
+				} else {
+					rects = [ start.getClientRect(), start.getClientRect() ];
+					start.remove();
+				}
+				rect.right = Math.max( rects[ 0 ].right, rects[ 1 ].right );
+				rect.bottom = Math.max( rects[ 0 ].bottom, rects[ 1 ].bottom );
+				rect.left = Math.min( rects[ 0 ].left, rects[ 1 ].left );
+				rect.top = Math.min( rects[ 0 ].top, rects[ 1 ].top );
+				rect.width = Math.abs( rects[ 0 ].left - rects[ 1 ].left );
```

Comment:
I am thinking if i should use extra tabulations to improve readability. i.e:
```
				rect.bottom	= Math.max( rects[ 0].bottom,	rects[ 1 ].bottom );
 				rect.left			= Math.min( rects[ 0 ].left,		rects[ 1 ].left );
 				rect.top			= Math.min( rects[ 0 ].top,		rects[ 1 ].top );
 				rect.width		= Math.abs( rects[ 0 ].left -		rects[ 1 ].left );
 				rect.height		= rect.bottom - rect.top;
```

Reply:
Late answer but we are not using this kind of whitespace alignment. Reason is that it is just hard to maintain as you change one  of these 5 lines, you need to modify 4 other.

It's fine the way it currently is 👍